dev/bench/odbc/main.cc

```diff
@@ -0,0 +1,186 @@
+#include <sql.h>
+#include <sqlext.h>
+#include <sqltypes.h>
+#include <stdio.h>
+#include <string.h>
+#include <chrono>
+#include <fstream>
+#include <iostream>
+#include <map>
+#include <string>
+#include <vector>
+
+#define ERRMSG_LEN 200
+
+SQLINTEGER checkError(SQLRETURN rc, SQLSMALLINT handleType, SQLHANDLE handle,
+                      SQLWCHAR* errmsg) {
+  SQLRETURN retcode = SQL_SUCCESS;
+
+  SQLSMALLINT errNum = 1;
+  SQLWCHAR sqlState[10];
+  SQLINTEGER nativeError;
+  SQLWCHAR errMsg[ERRMSG_LEN];
+  SQLSMALLINT textLengthPtr;
+
+  if ((rc != SQL_SUCCESS) && (rc != SQL_SUCCESS_WITH_INFO) && (rc != SQL_NO_DATA)) {
+    SQLLEN numRecs = 0;
+    SQLGetDiagField(SQL_HANDLE_STMT, handle, 0, SQL_DIAG_NUMBER, &numRecs, 0, 0);
+    while (retcode != SQL_NO_DATA) {
+      retcode = SQLGetDiagRecW(handleType, handle, errNum, sqlState, &nativeError, errMsg,
+                               ERRMSG_LEN, &textLengthPtr);
+
+      if (retcode == SQL_INVALID_HANDLE) {
+        std::cerr << "checkError function was called with an invalid handle!!"
+                  << std::endl;
+        return 1;
+      }
+
+      if ((retcode != SQL_SUCCESS) && (retcode != SQL_SUCCESS_WITH_INFO)) {
+        wprintf(L"ERROR: %d: %ls : %ls\n", nativeError, sqlState, errMsg);
+      }
+
+      errNum++;
+    }
+
+    wprintf(L"%ls\n", errmsg);
+    return 1;
+  }
+
+  return 0;
+}
+
+#define CHECK_OK(EXPR, ERROR)                                        \
+  do {                                                               \
+    auto ret = (EXPR);                                               \
+    if (checkError(ret, SQL_HANDLE_DBC, dbc, (SQLWCHAR*)L##ERROR)) { \
+      exit(1);                                                       \
+    }                                                                \
+  } while (0)
+
+int main(int argc, char** argv) {
+  SQLHENV env;
+  SQLHDBC dbc;
+  SQLHSTMT stmt;
+  SQLHSTMT stmt1;
+
+  std::string dsn(argv[0]);
+
+  SQLAllocHandle(SQL_HANDLE_ENV, SQL_NULL_HANDLE, &env);
+  // we want ODBC3 support, set env handle
+  SQLSetEnvAttr(env, SQL_ATTR_ODBC_VERSION, (void*)SQL_OV_ODBC3_80, 0);
+  // allocate connection handle
+  SQLAllocHandle(SQL_HANDLE_DBC, env, &dbc);
+
+  // connect to the DSN using dbc handle
+  CHECK_OK(SQLDriverConnect(dbc, nullptr, (SQLCHAR*)(dsn.c_str()), SQL_NTS, nullptr, 0,
```

Comment:
May need to define SQL_NOUNICODEMAP to prevent this from expecting SQLWCHAR* arguments. Or explicitly call SQLDriverConnectA() to call the ANSI function directly.